### PR TITLE
Improve safety from calls while in closed state

### PR DESCRIPTION
**Pull request by Mats-SX**
Also adding `toString` in two places for debugging purposes.

### CODE REVIEW

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala [hunk 1]
```diff
@@ -138,7 +138,7 @@ abstract class CompiledExecutionResult(taskCloser: TaskCloser, statement:Stateme
 
   override def notifications = Iterable.empty[InternalNotification]
 
-  override def hasNext = innerIterator.hasNext
+  override def hasNext = !taskCloser.isClosed && innerIterator.hasNext
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala]: This is not correct. The idea is to load the `innerIterator` on the first usage and then we expect statement to be closed. So basically `taskCloser.isClosed` should almost always be `true` here.

Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala]: Is the contract of `hasNext` such that calling it when a previous call returned `false` produces undefined behaviour? If not, we need this kind of check at some point between calling `hasNext` and entering `accept` method.

EDIT: Also, if `taskCloser.isClosed` almost always is true, then this is not a problem, right? True means `hasNext` evaluates to false, which is fine, and not slower than evaluating `innerIterator.hasNext`.

Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala]: No it is not fine if there are more items to iterate, we will almost always prematurely abort the iteration.

**Comment by Mats-SX** [comment]
@pontusmelke Do you think this solution is correct?

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResultTest.scala [hunk 2]
```diff
@@ -108,15 +109,38 @@ class CompiledExecutionResultTest extends CypherFunSuite {
     toList(result.javaIterator) should equal(javaList(javaMap("foo" -> javaMap("key" -> "value"))))
   }
 
-  private def newCompiledExecutionResult(row: util.Map[String, Any]) =
-    new CompiledExecutionResult(null, null, null, null) {
+  test("javaIterator hasNext should not call accept if results already consumed") {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResultTest.scala:114]: This test does not test what it claims to test.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala [hunk 3]
```diff
@@ -189,12 +189,14 @@ abstract class CompiledExecutionResult(taskCloser: TaskCloser, statement:Stateme
   }.mkString("{", ", ", "}")
 
   private def doInAccept[T](body: ResultRow => T) = {
-    accept(new ResultVisitor[RuntimeException] {
-      override def visit(row: ResultRow): Boolean = {
-        body(row)
-        true
-      }
-    })
+    if (!taskCloser.isClosed) {
```
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:200]: I think we still should throw an exception here, maybe something nicer than the `Statement has been closed` but I still think it makes sense to throw here.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResultTest.scala (hunk 2 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResultTest.scala:114]: I fixed this.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala (hunk 3 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:200]: Done.

**Pull request merged**